### PR TITLE
Continue parser

### PR DESCRIPTION
**Pull request by 0Lucifer0**

### CODE REVIEW

**Comment by erwan-joly** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- src/NosCore.Parser/Parsers/ItemParser.cs  116
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/NosCore/NosCore/pullRequest?prid=4456396&bid=15061093)

File: src/NosCore.Parser/Parsers/ItemParser.cs [hunk 1]
```diff
@@ -0,0 +1,643 @@
...
+                ElementType.Dark => equipmentType == EquipmentType.Sp ? chunk["DATA"][0][18] : chunk["DATA"][0][11],
+                _ => "0"
+            });
+        }
+
+        private void SetVehicles(List<ItemDto> items, Dictionary<byte, List<(short, short)>> vehicleDictionary)
+        {
+            foreach (var vehicle in vehicleDictionary)
+            {
+                var speed = vehicle.Key;
```
Comment (erwan-joly) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Remove the unused local variable 'speed'.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+using NosCore.Data.Enumerations.Items;
+using NosCore.Data.StaticEntities;
+using NosCore.Parser.Parsers.Generic;
+using Serilog;
+
+namespace NosCore.Parser.Parsers
+{
+    public class ItemParser
+    {
+        //  VNUM	{VNum}	{Price}
```
Comment (erwan-joly) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Replace all tab characters in this file by sequences of white-spaces.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.PathFinder.Gui/GuiWindow.cs [hunk 3]
```diff
@@ -121,20 +122,30 @@ protected override void OnUpdateFrame(FrameEventArgs e)
...
             base.OnUpdateFrame(e);
 
-            if (Keyboard[Key.Escape])
+            if (KeyPress(Key.Escape))
             {
                 Exit();
             }
         }
 
+        KeyboardState keyboardState, lastKeyboardState;
```
Comment (erwan-joly) [inline src/NosCore.PathFinder.Gui/GuiWindow.cs]: Codacy Issue found: [Declare 'lastKeyboardState' in a separate statement.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+                    {
+                        item.SecondaryElement = elementdic.ElementAt(1).Key;
+                    }
+                }
+            }
+            SetVehicles(items, new Dictionary<byte, List<(short, short)>>
+            {
+                { 20, new List<(short, short)> { (9054, 2368), (1906, 2368), (9055, 2370), (1907, 2370), (9058, 2406), (1965, 2406), (9065, 2411), (5008, 2411), (5238, 1817), (5226, 1817), (5240, 1819), (5228, 1819), (5234, 2522), (5236, 2524) } },
+                { 21, new List<(short, short)> { (9070, 2429), (5117, 2429), (9073, 2432), (5152, 2432), (5196, 2517), (9078, 2520), (5232, 2520), (5321, 2528), (9090, 2934), (5386, 2934), (9091, 2936), (5387, 2936), (9092, 2938), (5388, 2938), (9093, 2940), (5389, 2940),
+                    (9094, 2942), (5390, 2942), (5391, 2944), (9115, 3679), (5997, 3679), (9079, 2522), (9080, 2524), (9081, 1817), (9082, 1819) } },
```
Comment (erwan-joly) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Move this closing curly brace to the next line.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+                    item.Element = elementdic.First().Key;
+                    if (elementdic.First().Value != 0 && elementdic.First().Value == elementdic.ElementAt(1).Value)
+                    {
+                        item.SecondaryElement = elementdic.ElementAt(1).Key;
+                    }
+                }
+            }
+            SetVehicles(items, new Dictionary<byte, List<(short, short)>>
+            {
+                { 20, new List<(short, short)> { (9054, 2368), (1906, 2368), (9055, 2370), (1907, 2370), (9058, 2406), (1965, 2406), (9065, 2411), (5008, 2411), (5238, 1817), (5226, 1817), (5240, 1819), (5228, 1819), (5234, 2522), (5236, 2524) } },
```
Comment (erwan-joly) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Split this 248 characters long line (which is greater than 200 authorized).](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs [hunk 6]
```diff
@@ -0,0 +1,643 @@
...
+                { 21, new List<(short, short)> { (9070, 2429), (5117, 2429), (9073, 2432), (5152, 2432), (5196, 2517), (9078, 2520), (5232, 2520), (5321, 2528), (9090, 2934), (5386, 2934), (9091, 2936), (5387, 2936), (9092, 2938), (5388, 2938), (9093, 2940), (5389, 2940),
```
Comment (erwan-joly) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Split this 272 characters long line (which is greater than 200 authorized).](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs [hunk 7]
```diff
@@ -0,0 +1,643 @@
...
+                    Type = type,
+                    SubType = (byte)((int.Parse(chunk["BUFF"][i][5 + 5 * i]) + 1) * 10 + 1),
+                    IsLevelScaled = Convert.ToBoolean(first % 4),
+                    IsLevelDivided = (uint)(first < 0 ? 0 : first) % 4 == 2,
+                    FirstData = (short)(first > 0 ? first : -first / 4),
+                    SecondData = (short)(int.Parse(chunk["BUFF"][0][4 + 5 * i]) / 4),
+                    ThirdData = (short)(int.Parse(chunk["BUFF"][0][6 + 5 * i]) / 4)
+                };
+                list.Add(comb);
+            };
```
Comment (erwan-joly) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Remove this empty statement.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.PathFinder.Gui/GuiWindow.cs (hunk 3 above)
Comment (erwan-joly) [inline src/NosCore.PathFinder.Gui/GuiWindow.cs]: Codacy Issue found: [Remove unassigned field 'lastKeyboardState', or set its value.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

**Comment by 0Lucifer0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- src/NosCore.Parser/Parsers/ItemParser.cs  116
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/NosCore/NosCore/pullRequest?prid=4456396&bid=15061093)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 7 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Remove this empty statement.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 2 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Replace all tab characters in this file by sequences of white-spaces.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 4 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Move this closing curly brace to the next line.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.PathFinder.Gui/GuiWindow.cs (hunk 3 above)
Comment (0Lucifer0) [inline src/NosCore.PathFinder.Gui/GuiWindow.cs]: Codacy Issue found: [Remove unassigned field 'lastKeyboardState', or set its value.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 6 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Split this 272 characters long line (which is greater than 200 authorized).](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.PathFinder.Gui/GuiWindow.cs (hunk 3 above)
Comment (0Lucifer0) [inline src/NosCore.PathFinder.Gui/GuiWindow.cs]: Codacy Issue found: [Declare 'lastKeyboardState' in a separate statement.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 5 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Split this 248 characters long line (which is greater than 200 authorized).](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 1 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Remove the unused local variable 'speed'.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

**Comment by 0Lucifer0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- src/NosCore.Parser/Parsers/ItemParser.cs  116
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/NosCore/NosCore/pullRequest?prid=4456396&bid=15061093)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 2 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Replace all tab characters in this file by sequences of white-spaces.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 4 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Move this closing curly brace to the next line.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.PathFinder.Gui/GuiWindow.cs (hunk 3 above)
Comment (0Lucifer0) [inline src/NosCore.PathFinder.Gui/GuiWindow.cs]: Codacy Issue found: [Remove unassigned field 'lastKeyboardState', or set its value.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

Comment (0Lucifer0) [inline src/NosCore.PathFinder.Gui/GuiWindow.cs]: Codacy Issue found: [Declare 'lastKeyboardState' in a separate statement.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 1 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Remove the unused local variable 'speed'.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 7 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Remove this empty statement.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 6 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Split this 272 characters long line (which is greater than 200 authorized).](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 5 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Split this 248 characters long line (which is greater than 200 authorized).](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

**Comment by 0Lucifer0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 6
           

Complexity increasing per file
==============================
- src/NosCore.Parser/Parsers/ItemParser.cs  123
- src/NosCore.Parser/Parsers/Generic/GenericParser.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/NosCore/NosCore/pullRequest?prid=4456396&bid=15061093)

File: src/NosCore.Parser/Parsers/ItemParser.cs [hunk 8]
```diff
@@ -0,0 +1,679 @@
...
+using NosCore.Data.Enumerations.Items;
+using NosCore.Data.StaticEntities;
+using NosCore.Parser.Parsers.Generic;
+using Serilog;
+
+namespace NosCore.Parser.Parsers
+{
+    public class ItemParser
+    {
+        //  VNUM	{VNum}	{Price}
```
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Replace all tab characters in this file by sequences of white-spaces.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs [hunk 9]
```diff
@@ -0,0 +1,679 @@
...
+                    if (elementdic.First().Value != 0 && elementdic.First().Value == elementdic.ElementAt(1).Value)
+                    {
+                        item.SecondaryElement = elementdic.ElementAt(1).Key;
+                    }
+                }
+            }
+            SetVehicles(items, new Dictionary<byte, List<(short, short)>>
+            {
+                { 20, new List<(short, short)> { (9054, 2368), (1906, 2368), (9055, 2370), (1907, 2370), (9058, 2406), (1965, 2406), (9065, 2411),
+                    (5008, 2411), (5238, 1817), (5226, 1817), (5240, 1819), (5228, 1819), (5234, 2522), (5236, 2524) } },
```
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Move this closing curly brace to the next line.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs [hunk 10]
```diff
@@ -0,0 +1,679 @@
...
+                    Type = type,
+                    SubType = (byte)((int.Parse(chunk["BUFF"][0][5 + 5 * i]) + 1) * 10 + 1),
+                    IsLevelScaled = Convert.ToBoolean(first % 4),
+                    IsLevelDivided = (uint)(first < 0 ? 0 : first) % 4 == 2,
+                    FirstData = (short)(first > 0 ? first : -first / 4),
+                    SecondData = (short)(int.Parse(chunk["BUFF"][0][4 + 5 * i]) / 4),
+                    ThirdData = (short)(int.Parse(chunk["BUFF"][0][6 + 5 * i]) / 4)
+                };
+                list.Add(comb);
+            };
```
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Remove this empty statement.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs [hunk 11]
```diff
@@ -0,0 +1,679 @@
...
+                {nameof(ItemDto.VNum), chunk => Convert.ToInt16(chunk["VNUM"][0][2])},
+                {nameof(ItemDto.Price), chunk => Convert.ToInt64(chunk["VNUM"][0][3])},
+                {nameof(ItemDto.ReputPrice), chunk => chunk["FLAG"][0][21] == "1" ? Convert.ToInt64(chunk["VNUM"][0][3]) : 0},
+                {nameof(ItemDto.NameI18NKey), chunk => chunk["NAME"][0][2]},
+                {nameof(ItemDto.Type), chunk => ImportType(chunk)},
+                {nameof(ItemDto.ItemType), chunk => ImportItemType(chunk)},
+                {nameof(ItemDto.ItemSubType), chunk => Convert.ToByte(chunk["INDEX"][0][4])},
+                {nameof(ItemDto.EquipmentSlot), chunk => ImportEquipmentType(chunk)},
+                {nameof(ItemDto.Morph), chunk =>  ImportEffect(chunk) == ItemEffectType.ApplySkinPartner ?Convert.ToInt16(chunk["INDEX"][0][5]) :
+                    ImportEquipmentType(chunk) != EquipmentType.Amulet ? Convert.ToInt16(chunk["INDEX"][0][7]) : default},
```
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Move this closing curly brace to the next line.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/I18NParser.cs [hunk 12]
```diff
@@ -54,33 +54,34 @@ private string I18NTextFileName(string textfilename, RegionType region)
...
+                        if(line == null)
+                        {
+                            continue;
+                        }
+                        var currentLine = line.Split('\t');
+                        if (currentLine.Length > 1 && !listoftext.ContainsKey((currentLine[0], region))  &&
+                            !dtos.ContainsKey(currentLine[0]))
                         {
-                            dtos.Add(new TDto()
+                            dtos.Add(currentLine[0], new TDto()
```
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/I18NParser.cs]: Codacy Issue found: [Remove these redundant parentheses.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs [hunk 13]
```diff
@@ -0,0 +1,679 @@
...
+                    }
+                }
+            }
+            SetVehicles(items, new Dictionary<byte, List<(short, short)>>
+            {
+                { 20, new List<(short, short)> { (9054, 2368), (1906, 2368), (9055, 2370), (1907, 2370), (9058, 2406), (1965, 2406), (9065, 2411),
+                    (5008, 2411), (5238, 1817), (5226, 1817), (5240, 1819), (5228, 1819), (5234, 2522), (5236, 2524) } },
+                { 21, new List<(short, short)> { (9070, 2429), (5117, 2429), (9073, 2432), (5152, 2432), (5196, 2517), (9078, 2520), (5232, 2520),
+                    (5321, 2528), (9090, 2934), (5386, 2934), (9091, 2936), (5387, 2936), (9092, 2938), (5388, 2938), (9093, 2940), (5389, 2940),
+                    (9094, 2942), (5390, 2942), (5391, 2944), (9115, 3679), (5997, 3679), (9079, 2522), (9080, 2524), (9081, 1817), (9082, 1819) } },
```
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Move this closing curly brace to the next line.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

**Comment by 0Lucifer0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 6
           

Complexity increasing per file
==============================
- src/NosCore.Parser/Parsers/Generic/GenericParser.cs  1
- src/NosCore.Parser/Parsers/ItemParser.cs  123
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/NosCore/NosCore/pullRequest?prid=4456396&bid=15061093)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 13 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Move this closing curly brace to the next line.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 11 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Move this closing curly brace to the next line.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/I18NParser.cs (hunk 12 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/I18NParser.cs]: Codacy Issue found: [Remove these redundant parentheses.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 9 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Move this closing curly brace to the next line.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 10 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Remove this empty statement.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

File: src/NosCore.Parser/Parsers/ItemParser.cs (hunk 8 above)
Comment (0Lucifer0) [inline src/NosCore.Parser/Parsers/ItemParser.cs]: Codacy Issue found: [Replace all tab characters in this file by sequences of white-spaces.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4456396)

**Pull request merged**